File: python/71-80/Minimum Window Substring.py
    # @return a string
    def minWindow(self, S, T):
        if S=='' or T=='':
            return ''
        Dics={}
        temp={}
        for c in T:
            if  c in Dics:
                Dics[c]+=1
            else:
                Dics[c]=1
        counter,num=0,len(Dics)
        Min,result=len(S),''
        begin,end=0,0
        while end < len(S):
            ch = S[end]
            if ch in Dics:
                if ch in temp:
                    temp[ch]+=1
                else:
                    temp[ch]=1
                if temp[ch]==Dics[ch]:
                    counter+=1
                    
                if counter == num:
                    while counter==num:
                        if end-begin+1 <=Min:
                            Min = end-begin+1
                            result =S[begin:end+1]
                        c = S[begin]
                        begin+=1
                        if c in temp:
                            temp[c]-=1
                            if temp[c]<Dics[c]:
                                counter-=1
                        
            end+=1
        return result

File: Cpp/71-80/Minimum Window Substring.cpp
    /**
     *  runtime: 132ms
     *  error: too many
     * 
     */ 
       string minWindow(string s, string t) {
        if(s.size()==0||t.size()==0) return "";
        unordered_map<char,int> myMap;
        unordered_map<char,int> temp;
        for(auto ch: t)
            myMap[ch]++;
        string res="";
        int num = 0;
        int beg=0,end=0;
        while(end<s.size()){
            char ch = s[end];
            if(myMap.find(ch)==myMap.end()){
                end++;
                continue;
            }
            temp[ch]++;
            if(temp[ch]==myMap[ch])
                num++;
            while(num==myMap.size()){
                // if(res==""||res.size()>end-beg+1)
                //     res = s.substr(beg,end-beg+1);
                if(myMap.find(s[beg])!=myMap.end()){
                    if(temp[s[beg]]==myMap[s[beg]])
                        num--;
                    temp[s[beg]]--;
                }
                if(res==""||res.size()>end-beg+1)
                res = s.substr(beg,end-beg+1);
                beg++;
            }
            end++;
        }return res;
    }  
